File: preprocess/preprocess_streamflow_data.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 27 12:26:33 2025

@author: Diego Dinamarca
"""

import pandas as pd
from datetime import datetime
import os
import json

# Load configuration from JSON
script_dir = os.path.dirname(os.path.abspath(__file__))
config_path = os.path.join(script_dir, "preprocess_config.json")
with open(config_path, "r") as f:
    config = json.load(f)

streamflow_data_file = config["streamflow_data_file"]
gauge_list = config["gauge_list"]
gauge_folder = config["gauge_folder"]


# Read CSV assuming the first column is the date
df = pd.read_csv(streamflow_data_file, parse_dates=[0])
df.set_index(df.columns[0], inplace=True)

for i, gauge_id in enumerate(gauge_list, start=1):
    if gauge_id not in df.columns:
        print(f"Gauge ID {gauge_id} not found in data.")
        continue

    series_full = df[gauge_id]
    series_valid = series_full.dropna()

    if series_valid.empty:
        print(f"No valid data for gauge {gauge_id}. Skipping.")
        continue

    start_date = series_valid.index.min()
    end_date = series_valid.index.max()

    # Create complete daily time index
    full_range = pd.date_range(start=start_date, end=end_date, freq='D')
    filled_series = series_full.reindex(full_range).fillna(-9999)

    # Header lines
    output_lines = [
        f"{gauge_id} Gauge {i} (daily discharge)",
        "nodata   -9999",
        "n       1       measurements per day [1, 1440]",
        f"start  {start_date:%Y %m %d %H %M}",
        f"end    {end_date:%Y %m %d %H %M}"
    ]

    # Data lines
    for date, value in filled_series.items():
        line = f"{date:%Y  %m  %d  %H  %M}   {value:8.3f}"
        output_lines.append(line)

    output_file = os.path.join(gauge_folder, f"{gauge_id}.txt")
    with open(output_file, 'w') as f:
        f.write("\n".join(output_lines))

    print(f"Wrote file: {output_file}")

File: R/preprocess_streamflow_data.R
library(jsonlite)
library(lubridate)
library(tidyverse)

# === Load configuration ===
config <- fromJSON("/Users/mhm/Desktop/FONDECYT_CAMILA/mhm_snow/SCRIPTS/preprocess_config.json")

streamflow_data_file <- config$streamflow_data_file
gauge_list <- config$gauge_list
gauge_folder <- config$gauge_folder

dir.create(gauge_folder, showWarnings = FALSE, recursive = TRUE)

# === Read CSV assuming the first column is date ===
df <- read_csv(streamflow_data_file)
colnames(df)[1] <- "Date"

for (i in seq_along(gauge_list)) {
  gauge_id <- gauge_list[[i]]
  if (!(gauge_id %in% colnames(df))) {
    cat(sprintf("Gauge ID %s not found in data.\n", gauge_id))
    next
  }
  
  series_valid <- df %>% select(Date, !!gauge_id) %>% drop_na()
  series_valid

  if (nrow(series_valid) == 0) {
    cat(sprintf("No valid data for gauge %s. Skipping.\n", gauge_id))
    next
  }
  
  start_date <- min(series_valid$Date)
  end_date <- max(series_valid$Date)
  
  
  full_range <- data.frame(Date = seq(start_date, end_date, by = "1 day"))
  filled_series <- full_range %>%
    left_join(series_valid, by = "Date") %>%
    mutate(Value = ifelse(is.na(.data[[gauge_id]]), -9999, .data[[gauge_id]])) %>%
    mutate(Hour = 0, Minute = 0)
  
  # Header lines
  output_lines <- c(
    sprintf("%s Gauge %d (daily discharge)", gauge_id, i),
    "nodata   -9999",
    "n       1       measurements per day [1, 1440]",
    sprintf("start  %s", format(start_date, "%Y %m %d %H %M")),
    sprintf("end    %s", format(end_date, "%Y %m %d %H %M"))
  )
  
  # Data lines
  data_lines <- filled_series %>%
    mutate(Line = sprintf("%4d  %02d  %02d  %02d  %02d   %8.3f",
                          year(Date), month(Date), day(Date), Hour, Minute, Value)) %>%
    pull(Line)
  
  output_lines <- c(output_lines, data_lines)
  
  # Write output
  output_file <- file.path(gauge_folder, paste0(gauge_id, ".txt"))
  writeLines(output_lines, output_file)
  
  cat(sprintf("Wrote file: %s\n", output_file))
}
